File: c/3-pid-bm/main/wifi.c
/* WiFi station Example

   This example code is in the Public Domain (or CC0 licensed, at your option.)

   Unless required by applicable law or agreed to in writing, this
   software is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied.
*/
#include <string.h>

#include "esp_err.h"
#include "esp_event.h"
#include "esp_log.h"
#include "esp_wifi.h"
#include "freertos/event_groups.h"
#include "freertos/task.h"

#include "wifi.h"

#define ESP_WIFI_SSID CONFIG_ESP_WIFI_SSID
#define ESP_WIFI_PASS CONFIG_ESP_WIFI_PASSWORD
#define ESP_MAXIMUM_RETRY CONFIG_ESP_MAXIMUM_RETRY

/* The event group allows multiple bits for each event, but we only care about
 * two events:
 * - we are connected to the AP with an IP
 * - we failed to connect after the maximum amount of retries */
#define WIFI_CONNECTED_BIT BIT0
#define WIFI_FAIL_BIT BIT1

static const char *TAG = "wifi";

/* FreeRTOS event group to signal when we are connected*/
static EventGroupHandle_t s_wifi_event_group;
static int s_retry_num = 0;

static void event_handler(
    void *arg, esp_event_base_t event_base, int32_t event_id, void *event_data
) {
  if (event_base == WIFI_EVENT && event_id == WIFI_EVENT_STA_START) {
    esp_wifi_connect();
  } else if (event_base == WIFI_EVENT &&
             event_id == WIFI_EVENT_STA_DISCONNECTED) {
    if (s_retry_num < ESP_MAXIMUM_RETRY) {
      esp_wifi_connect();
      s_retry_num++;
      ESP_LOGI(TAG, "retry to connect to the AP");
    } else {
      xEventGroupSetBits(s_wifi_event_group, WIFI_FAIL_BIT);
    }
    ESP_LOGI(TAG, "connect to the AP fail");
  } else if (event_base == IP_EVENT && event_id == IP_EVENT_STA_GOT_IP) {
    ip_event_got_ip_t *event = (ip_event_got_ip_t *)event_data;
    ESP_LOGI(TAG, "got ip:" IPSTR, IP2STR(&event->ip_info.ip));
    s_retry_num = 0;
    xEventGroupSetBits(s_wifi_event_group, WIFI_CONNECTED_BIT);
  }
}

esp_err_t my_wifi_init(my_wifi_t *self) {
  esp_err_t err;

  // Init
  if (s_wifi_event_group != NULL) {
    ESP_LOGE(TAG, "Wifi already set up");
    return ESP_ERR_INVALID_STATE;
  }
  s_wifi_event_group = xEventGroupCreate();
  if (s_wifi_event_group == NULL) {
    ESP_LOGE(TAG, "xEventGroupCreate fail");
    return ESP_ERR_INVALID_STATE;
  }

  err = esp_netif_init();
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_netif_init fail (0x%x)", (int)err);
    return err;
  }

  esp_netif_t *netif = esp_netif_create_default_wifi_sta();
  if (netif == NULL) {
    ESP_LOGE(TAG, "esp_netif_create_default_wifi_sta fail (0x%x)", (int)err);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return ESP_ERR_INVALID_STATE;
  }

  wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
  err = esp_wifi_init(&cfg);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_wifi_init fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }

  // Handlers
  esp_event_handler_instance_t handler_wifi;
  err = esp_event_handler_instance_register(
      WIFI_EVENT, ESP_EVENT_ANY_ID, &event_handler, NULL, &handler_wifi
  );
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_event_handler_instance_register fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }

  esp_event_handler_instance_t handler_ip;
  err = esp_event_handler_instance_register(
      IP_EVENT, IP_EVENT_STA_GOT_IP, &event_handler, NULL, &handler_ip
  );
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_event_handler_instance_register fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }

  // Config
  err = esp_wifi_set_mode(WIFI_MODE_STA);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_wifi_set_mode fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }
  wifi_config_t wifi_config = {
      .sta =
          {
              .ssid = ESP_WIFI_SSID,
              .password = ESP_WIFI_PASS,
              .threshold.authmode = WIFI_AUTH_WPA2_PSK,
              .sae_pwe_h2e = WPA3_SAE_PWE_UNSPECIFIED,
              .sae_h2e_identifier = "",
          },
  };
  err = esp_wifi_set_config(WIFI_IF_STA, &wifi_config);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_wifi_set_config fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }
  err = esp_wifi_start();
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_wifi_start fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }

  /* Waiting until either the connection is established (WIFI_CONNECTED_BIT) or
   * connection failed for the maximum number of re-tries (WIFI_FAIL_BIT). The
   * bits are set by event_handler() (see above) */
  EventBits_t bits = xEventGroupWaitBits(
      s_wifi_event_group, WIFI_CONNECTED_BIT | WIFI_FAIL_BIT, pdFALSE, pdFALSE,
      portMAX_DELAY
  );

  /* xEventGroupWaitBits() returns the bits before the call returned, hence we
   * can test which event actually happened. */
  if (bits & WIFI_CONNECTED_BIT) {
    ESP_LOGI(TAG, "connected to ap SSID:%s", ESP_WIFI_SSID);
  } else if (bits & WIFI_FAIL_BIT) {
    ESP_LOGI(TAG, "Failed to connect to SSID:%s", ESP_WIFI_SSID);
  } else {
    ESP_LOGE(TAG, "UNEXPECTED EVENT");
  }

  err = esp_wifi_set_ps(WIFI_PS_NONE);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "esp_wifi_start fail (0x%x)", (int)err);
    esp_netif_destroy_default_wifi(netif);
    ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
    return err;
  }

  *self = (my_wifi_t){.netif = netif};

  return ESP_OK;
}

void my_wifi_deinit(my_wifi_t *self) {
  esp_netif_destroy_default_wifi(self->netif);
  ESP_ERROR_CHECK_WITHOUT_ABORT(esp_netif_deinit());
}

File: zig/3-pid-bm/main/Wifi.zig
const std = @import("std");
const idf = @import("esp_idf");
const sys = idf.sys;

const c = @import("c.zig");
const logErr = @import("utils.zig").logErr;

const ssid = c.CONFIG_ESP_WIFI_SSID;
const password = c.CONFIG_ESP_WIFI_PASSWORD;
const maximum_retries = 5;

const Self = @This();

pub const EventFlags = packed struct(sys.EventBits_t) {
    failed: bool = false,
    connected: bool = false,
    _padding: u30 = 0,

    fn all() EventFlags {
        return .{ .failed = true, .connected = true };
    }
};

var event_group: sys.EventGroupHandle_t = null;
var retry_count: u32 = 0;

netif: *sys.esp_netif_t,

pub fn init() !Self {
    if (event_group) |_| return error.ErrorWifiAlreadySetUp;

    const event_group_local = sys.xEventGroupCreate() orelse return error.ErrorSetupWifi;
    event_group = event_group_local;

    try idf.espCheckError(sys.esp_netif_init());
    errdefer idf.espLogError(sys.esp_netif_deinit(), "esp_netif_deinit");

    const netif = sys.esp_netif_create_default_wifi_sta() orelse return error.ErrorSetupWifi;
    errdefer sys.esp_netif_destroy_default_wifi(netif);

    try idf.wifi.init(&c.fixed.wifiInitConfigDefault());
    errdefer logErr(idf.wifi.deinit(), "Wifi.deinit");

    var handler_wifi: sys.esp_event_handler_instance_t = undefined;
    try idf.espCheckError(sys.esp_event_handler_instance_register(
        sys.WIFI_EVENT,
        c.ESP_EVENT_ANY_ID,
        &eventHandler,
        null,
        &handler_wifi,
    ));

    var handler_ip: sys.esp_event_handler_instance_t = undefined;
    try idf.espCheckError(sys.esp_event_handler_instance_register(
        sys.IP_EVENT,
        @intFromEnum(sys.ip_event_t.IP_EVENT_STA_GOT_IP),
        &eventHandler,
        null,
        &handler_ip,
    ));

    try idf.wifi.setMode(.WIFI_MODE_STA);

    var wifi_config = sys.wifi_config_t{
        .sta = .{
            .ssid = nullExtend(@FieldType(sys.wifi_sta_config_t, "ssid"), ssid),
            .password = nullExtend(@FieldType(sys.wifi_sta_config_t, "password"), password),
            .threshold = .{ .authmode = sys.wifi_auth_mode_t.WIFI_AUTH_WPA2_PSK },
        },
    };
    try idf.wifi.setConfig(.WIFI_IF_STA, &wifi_config);

    try idf.wifi.start();
    errdefer logErr(idf.wifi.stop(), "Wifi.stop");

    const bits = sys.xEventGroupWaitBits(
        event_group_local,
        @bitCast(EventFlags.all()),
        @intFromBool(false),
        @intFromBool(false),
        std.math.maxInt(sys.TickType_t),
    );
    const flags: EventFlags = @bitCast(bits);
    if (flags.connected) {
        std.log.info("Connected to AP SSID:{s}", .{ssid});
    } else if (flags.failed) {
        return error.ErrorWifiConnectionFailed;
    }

    try idf.wifi.PowerSave.set(.WIFI_PS_NONE);

    return .{ .netif = netif };
}

pub fn deinit(self: *const Self) void {
    event_group = null;
    logErr(idf.wifi.stop(), "Wifi.stop");
    logErr(idf.wifi.deinit(), "Wifi.deinit");
    sys.esp_netif_destroy_default_wifi(self.netif);
    idf.espLogError(sys.esp_netif_deinit(), "esp_netif_deinit");
}

fn eventHandler(
    args: ?*anyopaque,
    event_base: sys.esp_event_base_t,
    event_id: i32,
    event_data: ?*anyopaque,
) callconv(.c) void {
    eventHandlerImpl(args, event_base, event_id, event_data) catch |err| {
        std.log.err("Wifi.eventHandler fail: {}", .{err});
    };
}

fn eventHandlerImpl(
    _: ?*anyopaque,
    event_base: sys.esp_event_base_t,
    event_id: i32,
    event_data: ?*anyopaque,
) !void {
    const event_group_local = event_group orelse {
        return error.WifiNotInitialized;
    };
    const event = EventType{ .base = event_base, .id = @bitCast(event_id) };

    if (event.is(&EventType.wifi(.WIFI_EVENT_STA_START))) {
        try idf.wifi.connect();
    } else if (event.is(&EventType.wifi(.WIFI_EVENT_STA_DISCONNECTED))) {
        if (retry_count < maximum_retries) {
            try idf.wifi.connect();
            retry_count += 1;
            std.log.info(
                "Retrying to connect to the AP ({}/{})",
                .{ retry_count, maximum_retries },
            );
        } else {
            std.log.err("Connecting to the AP failed", .{});
            _ = sys.xEventGroupSetBits(
                event_group_local,
                @bitCast(EventFlags{ .failed = true }),
            );
        }
    } else if (event.is(&EventType.ip(.IP_EVENT_STA_GOT_IP))) {
        const data: *sys.ip_event_got_ip_t = @alignCast(@ptrCast(event_data));
        const addr: [4]u8 = @bitCast(data.ip_info.ip.addr);
        std.log.info("Got ip: {}.{}.{}.{}", .{ addr[0], addr[1], addr[2], addr[3] });
        retry_count = 0;
        _ = sys.xEventGroupSetBits(
            event_group_local,
            @bitCast(EventFlags{ .connected = true }),
        );
    }
}

const EventType = struct {
    base: sys.esp_event_base_t,
    id: c_uint,

    pub fn wifi(id: sys.wifi_event_t) EventType {
        return .{ .base = sys.WIFI_EVENT, .id = @intFromEnum(id) };
    }
    pub fn ip(id: sys.ip_event_t) EventType {
        return .{ .base = sys.IP_EVENT, .id = @intFromEnum(id) };
    }

    pub fn is(self: *const EventType, other: *const EventType) bool {
        return std.mem.eql(
            u8,
            @ptrCast(self[0..1]),
            @ptrCast(other[0..1]),
        );
    }
};

fn nullExtend(comptime T: type, comptime str: [:0]const u8) [@sizeOf(T)]u8 {
    return (str ++ (.{'\x00'} ** (@sizeOf(T) - str.len))).*;
}
